### ISSUE TITLE
Customize Array Aggregate Display

### ISSUE THREAD

**Issue by RFSH** [informatics-isi-edu/ermrestjs#699]
Currently for pseudo-columns with `aggregate: "array"` we're returning a string of comma separated values. In entity mode, these values are links to the row with the row-name as the caption.

A new requirement is showing a bullet list of values instead of comma separated string. These are the potential solutions to this:

1. Adding a new attribute to the pseudo-columns that switched between comma separated string, and bullet list. What should we name this? We have `ux_mode`, which is only used in faceting. We could use the same attribute but with different set of acceptable values, but that is going to be more confusing. 

2. Letting DBA to define `prefix`, `postfix`, and `separator` that should be used for generating this value. This will be consistent with `row_markdown_pattern` in [table display annotation](https://github.com/informatics-isi-edu/ermrest/blob/master/docs/user-doc/annotation.md#2016-table-display). This will give more freedom to the DBA, but will pollute the pseudo-column definition. Naming of these attributes is another problem. Their name must be very specific that this is going to be used only for array aggregate.

3. Use the `prefix`, `postfix`, and `separator` that are defined in the table display annotation when the pseudo-column is in entity mode. This makes the pseudo-column definition simpler, but what should we do about the scalar pseudo-columns then?

The first two solutions require adding additional attributes to the pseudo-column definition. The following is the list of attributes that we have in faceting and pseudo-columns:

Required attributes:
- `source`

Common attributes: 
 - `entity`
 - `markdown_name`: displayname.
 - `comment`: tooltip.

Facet specific attributes:
- `choices`, `ranges`,  `not_null`: to define a pre-selected facet.
- `ux_mode`: facet UX mode (`choices`, `ranges`, `check_presence`)
- `bar_plot`: Histogram display settings.
- `open`: Setting this attribute to true, will force the facet to open by default.
- `order`:  Changes the default ordering of facet scalar values.
- `hide_num_occurrences`: Hide "number of occurrences" column from facet modal.

Pseudo-Column specific attributes:
- `aggregate`

**Comment by RFSH** [comment]
We decided to go with the first approach. We are going to introduce a new attribute for now that switched between comma separated string, and a list. If deployments don't like the bullet list, they can later remove that using css.

We didn't go with the other options because we didn't want to over complicate things. Besides in #617, we want to give DBAs the ability to use pseudo-columns in mustache/handlebars environment, which would be the better and more complete solution to this.

**Comment by carlkesselman** [comment]
what attribute name will you use?


----------------------------------------------------------
Dr. Carl Kesselman
Dean’s Professor, Epstein Department of Industrial and Systems Engineering
Fellow, Information Sciences Institute
Viterbi School of Engineering

Professor, Preventive Medicine
Keck School of Medicine

University of Southern California
4676 Admiralty Way, Suite 1001, Marina del Rey, CA  90292-6695
Phone:  +1 (310) 448-9338
Email:  carl@isi.edu<mailto:carl@isi.edu>
Web: http://www.isi.edu/~carl








On Aug 30, 2018, at 3:44 PM, Aref Shafaei <notifications@github.com<mailto:notifications@github.com>> wrote:


We decided to go with the first approach. We are going to introduce a new attribute for now that switched between comma separated string, and a list. If deployments don't like the bullet list, they can later remove that using css.

We didn't go with the other options because we didn't want to over complicate things. Besides in #617<https://github.com/informatics-isi-edu/ermrestjs/issues/617>, we want to give DBAs the ability to use pseudo-columns in mustache/handlebars environment, which would be the better and more complete solution to this.

**Comment by RFSH** [comment]
@carlkesselman I think I'm going to use `array_aggregate_display` which can be `list` or `csv` (by default it will be `csv`). I don't like the name, but I want it to be very explicit. Any other suggestions for it would be appreciated.

**Comment by carlkesselman** [comment]
is it an aggregate (which to me implies sum, max, min?).  How about just array_display: list, bullets
----------------------------------------------------------
Dr. Carl Kesselman
Dean’s Professor, Epstein Department of Industrial and Systems Engineering
Fellow, Information Sciences Institute
Viterbi School of Engineering

Professor, Preventive Medicine
Keck School of Medicine

University of Southern California
4676 Admiralty Way, Suite 1001, Marina del Rey, CA  90292-6695
Phone:  +1 (310) 448-9338
Email:  carl@isi.edu<mailto:carl@isi.edu>
Web: http://www.isi.edu/~carl








On Aug 30, 2018, at 4:27 PM, Aref Shafaei <notifications@github.com<mailto:notifications@github.com>> wrote:


@carlkesselman<https://github.com/carlkesselman> I think I'm going to use array_aggregate_display which can be list or csv (by default it will be csv). I don't like the name, but I want it to be very explicit. Any other suggestions for it would be appreciated.

**Comment by karlcz** [comment]
Yes it maps directly to the ermrest array aggregate interface which
maps to the postgres array_agg aggregate function (with an implicit
row_to_json() thrown in for entity-mode sources)...

**Comment by hongsudt** [comment]
@karlcz @robes are you ok with @carlkesselman's suggestion.. Using `list` for csv and 'bullet` for bullet list?

**Comment by robes** [comment]
@hongsudt I think I prefer Aref's original suggestion: `csv` or `list`. My second choice would be `csv` or `bulleted`. But I don't have a strong opinion.

**Comment by RFSH** [comment]
We decided to use `csv`, `ulist` (unordered list), and `olist` (ordered list) as the values for the `array_display`. I will update the annotation document when the PR is merged.

**Comment by RFSH** [comment]
This is merged with master. I also updated [the annotation document](https://github.com/informatics-isi-edu/ermrest/blob/master/docs/user-doc/annotation.md#2016-visible-columns).

**Comment by robes** [comment]
@hongsudt @RFSH weren't we also going to support a `raw` option for `array_display` so that we just render things without adding any formatting or separators. For example, to support a list of thumbnails?

Or do you want us to use it with one of these options and see how it looks first? I would try with `ulist` for example.

**Comment by RFSH** [comment]
I can add that. But just to be clear, array aggregate is not going to apply the `markdown_pattern` that you have defined for the column. So unless your column is actually a `markdown` that has the thumbnail image defined in it, you won't be able to use array aggregate (You can not create an array aggregate pseudo-column of a `url` field and expect it to apply the `markdown_pattern` that turns the url into image).

**Comment by robes** [comment]
Right. That's understood. The idea that we had in mind was to use the row_name annotation (on the _other_ entity) with markdown to generate an image and then aggregate that entity and display with a `raw` style so that in theory it will end up as a set of thumbnails.

**Comment by hongsudt** [comment]
@robes please try it with the existing display first.. If that works, we can do the raw support to remove `,`

**Comment by robes** [comment]
OK. We will do that.
